### PR TITLE
Can read project version from a file.

### LINKED ISSUE #7890: Simplify version extraction from a file
Providing some way to replace the following pattern would be welcome:

```meson
project('foo', 'c',
  version: run_command(
    find_program('python'), '-c', 'import sys;print(open(sys.argv[1]).read())', files('VERSION')
  ).stdout().strip(),
)
```

Drawbacks with this current approach:
- we can't move the `run_command()` in a variable before `project()`, which likely implies this is currently a nasty hack where we're probably not supposed to do that
- some system may have `python3` and not `python`, and the other way around is also true
- using `run_command(find_program('cat'), ...)` instead of `python` would be fatal on Windows
- having a dedicated Python script is quite an overhead, and doesn't address the external subprocess hack to just read a file

**Rationale**

So the current scenario we're in is the following: we are managing our releases through Github, where making a release is basically simply pushing a tag. So we have the following primitive `make-release.sh` script:

```sh
#!/bin/sh

#
# Release process:
# 1. on a clean git state, run this script with the new version as argument
# 2. check the last commit and tag
# 3. git push && git push --tags
#

set -eu

if [ $# -ne 1 ]; then
	echo "Usage $0 <major.minor.micro>"
	exit 1
fi

cd "$(dirname $0)"

if ! git diff-index --quiet HEAD; then
	echo "Git index is not clean"
	exit 1
fi

set -x
VERSION="$1"
echo "$VERSION" > VERSION
git add VERSION
git commit -m "Release $VERSION"
git tag "v$VERSION"
```

The `VERSION` file is basically the channel of communication between this script and `meson.build`

**Alternative considered**

- hot-patching `meson.build` within the release script, but it's not exactly cleaner to say the least
- using `dist` scripts with `meson`; unfortunately we're running git commands at an unfortunate timing, and it involves quite some logistic
- requesting in the release process to patch the `meson.build` with the new version, and then extracting the version with the meson introspection with tools like `jq` in the release script; this creates an annoying dependency to `jq`, and it also makes the release process much more clumsy than just running a script

**How it could look like**

Here are a two usage suggestions that would simplify and make this workflow much more reliable, which I believe are not very intrusive:

- `project('foo', 'c', version: files('VERSION'))` (it's part of the source files after all)
- `project('foo', 'c', version: meson.get_file_content('VERSION'))`

### PR DESCRIPTION
**Pull request by jpakkane**
Fixes #7890

Does not close #7931 as it is about reading file contents directly.

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -3169,9 +3169,24 @@ def func_project(self, node, args, kwargs):
...
+            if len(version) != 1:
+                raise InvalidCode('Version argument is an array with more than one entry.')
+            version = version[0]
+        if isinstance(version, mesonlib.File):
+            self.add_build_def_file(version)
+            ifname = version.absolute_path(self.environment.source_dir,
+                                           self.environment.build_dir)
+            try:
+                with open(ifname, 'r') as versionfile:
+                    self.project_version = versionfile.readline().strip()
```
Comment (eli-schwartz) [inline mesonbuild/interpreter.py:3183]: Should we be allowing version files that have additional non-version lines in it? Yes, it's documented as such, but there does not seem to be a good reason to allow it, and IMHO it is confusing.

Comment (jpakkane) [reply mesonbuild/interpreter.py:3183]: Either one is fine by me.

Comment (dcbaker) [reply mesonbuild/interpreter.py:3183]: For mesa our version file is: `21.0.0-rc4\n`. I'd prefer to error if there's more than one line, we can always change it later if there's a compelling reason, and it's a pretty trivial change for a project if they have more than one line.

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -3169,9 +3169,24 @@ def func_project(self, node, args, kwargs):
...
         self.active_projectname = proj_name
-        self.project_version = kwargs.get('version', 'undefined')
-        if not isinstance(self.project_version, str):
-            raise InvalidCode('The version keyword argument must be a string.')
+        version = kwargs.get('version', 'undefined')
+        if isinstance(version, list):
+            if len(version) != 1:
+                raise InvalidCode('Version argument is an array with more than one entry.')
+            version = version[0]
+        if isinstance(version, mesonlib.File):
```
Comment (dcbaker) [inline mesonbuild/interpreter.py:3177]: `elif`?

Comment (jpakkane) [reply mesonbuild/interpreter.py:3177]: Nope. That if just unpacks the item from the array. It is processed later. (`files` returns an array of files)

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -3169,9 +3169,28 @@ def func_project(self, node, args, kwargs):
...
+            self.add_build_def_file(version)
+            ifname = version.absolute_path(self.environment.source_dir,
+                                           self.environment.build_dir)
+            try:
+                ver_data = Path(ifname).read_text(encoding='utf-8').split('\n')
+            except FileNotFoundError:
+                raise InterpreterException('Version file not found.')
+            if len(ver_data) == 2 and ver_data[1] == '':
+                ver_data = ver_data[0:1]
```
Comment (eli-schwartz) [inline mesonbuild/interpreter.py:3186]: I would just `.rstrip('\n').split('\n')`, verify that it's exactly len() 1, and use that.

File: mesonbuild/interpreter.py (hunk 2 above)
Comment (dcbaker) [reply mesonbuild/interpreter.py:3177]: Oh, yeah, I see.

File: mesonbuild/interpreter.py (hunk 3 above)
Comment (jpakkane) [reply mesonbuild/interpreter.py:3186]: That accepts files that have multiple newlines at the end, which arguably should be rejected.

Comment (xclaesse) [reply mesonbuild/interpreter.py:3186]: If your line ends with `\r\n` I think the `\r` would be included in the version string, no?

Comment (xclaesse) [reply mesonbuild/interpreter.py:3186]:
```
>>> print('a\r\nb'.split('\n'))
['a\r', 'b']
>>> print('a\r\nb'.splitlines())
['a', 'b']
```

Comment (eli-schwartz) [reply mesonbuild/interpreter.py:3186]: read_text() uses universal_newlines

File: mesonbuild/interpreter.py [hunk 4]
```diff
@@ -3169,9 +3169,28 @@ def func_project(self, node, args, kwargs):
...
         self.active_projectname = proj_name
-        self.project_version = kwargs.get('version', 'undefined')
-        if not isinstance(self.project_version, str):
-            raise InvalidCode('The version keyword argument must be a string.')
+        version = kwargs.get('version', 'undefined')
+        if isinstance(version, list):
+            if len(version) != 1:
+                raise InvalidCode('Version argument is an array with more than one entry.')
+            version = version[0]
+        if isinstance(version, mesonlib.File):
```
Comment (xclaesse) [inline mesonbuild/interpreter.py:3177]: Missing FeatureNew()

File: mesonbuild/interpreter.py (hunk 3 above)
Comment (xclaesse) [reply mesonbuild/interpreter.py:3186]: oh didn't knowm good.

Comment (eli-schwartz) [reply mesonbuild/interpreter.py:3186]: https://docs.python.org/3/library/functions.html#open-newline-parameter

File: mesonbuild/interpreter.py (hunk 4 above)
Comment (jpakkane) [reply mesonbuild/interpreter.py:3177]: Added. This is the first one of these that I have added, hopefully I got it right.

Comment (dcbaker) [reply mesonbuild/interpreter.py:3177]: looks good.

**Pull request merged**